File: packages/constants/src/scoring.ts
import type { BurnoutRiskLevel, MBIDimension } from "@mindcheck/types";

export interface DimensionCutoff {
  dimension: MBIDimension;
  lowThreshold: number;
  highThreshold: number;
}

export interface DimensionCutoffMap {
  exhaustion: { low: number; high: number };
  cynicism: { low: number; high: number };
  efficacy: { low: number; high: number };
}

/**
 * Classify a single dimension score against its population cutoffs.
 * Efficacy is reverse-scored: lower efficacy = higher burnout risk.
 */
export function classifyDimension(
  value: number,
  dimension: MBIDimension,
  cutoffs: DimensionCutoffMap,
): "low" | "medium" | "high" {
  const c = cutoffs[dimension];

  if (dimension === "efficacy") {
    if (value >= c.high) return "low";
    if (value <= c.low) return "high";
    return "medium";
  }

  if (value <= c.low) return "low";
  if (value >= c.high) return "high";
  return "medium";
}

/**
 * Combine three dimension classifications into an overall risk level.
 * Two or more 'high' dimensions → high risk.
 * One 'high' or any 'medium' → medium risk.
 * All 'low' → low risk.
 */
export function overallRisk(
  exhaustion: number,
  cynicism: number,
  efficacy: number,
  cutoffs: DimensionCutoffMap,
): BurnoutRiskLevel {
  const levels = [
    classifyDimension(exhaustion, "exhaustion", cutoffs),
    classifyDimension(cynicism, "cynicism", cutoffs),
    classifyDimension(efficacy, "efficacy", cutoffs),
  ];

  const highs = levels.filter((l) => l === "high").length;
  if (highs >= 2) return "high";
  if (highs === 1 || levels.includes("medium")) return "medium";
  return "low";
}

/**
 * Build a DimensionCutoffMap from rows fetched from the
 * `instrument_cutoffs` table (one row per dimension).
 */
export function cutoffsFromRows(
  rows: Array<{ dimension: string; low_threshold: number; high_threshold: number }>,
): DimensionCutoffMap {
  const map: Partial<DimensionCutoffMap> = {};
  for (const row of rows) {
    if (row.dimension === "exhaustion" || row.dimension === "cynicism" || row.dimension === "efficacy") {
      map[row.dimension] = { low: row.low_threshold, high: row.high_threshold };
    }
  }
  if (!map.exhaustion || !map.cynicism || !map.efficacy) {
    throw new Error("Faltan cutoffs para alguna dimensión del MBI-SS");
  }
  return map as DimensionCutoffMap;
}

/**
 * Compute the mean score per MBI-SS dimension from raw item responses.
 * Item-to-dimension mapping comes from the `instrument_items` table.
 */
export function computeDimensionMeans(
  responses: Array<{ item_number: number; value: number }>,
  itemDimensions: Map<number, MBIDimension>,
): { exhaustion: number; cynicism: number; efficacy: number } {
  const sums = { exhaustion: 0, cynicism: 0, efficacy: 0 };
  const counts = { exhaustion: 0, cynicism: 0, efficacy: 0 };

  for (const r of responses) {
    const dim = itemDimensions.get(r.item_number);
    if (!dim) continue;
    sums[dim] += r.value;
    counts[dim] += 1;
  }

  return {
    exhaustion: counts.exhaustion ? sums.exhaustion / counts.exhaustion : 0,
    cynicism: counts.cynicism ? sums.cynicism / counts.cynicism : 0,
    efficacy: counts.efficacy ? sums.efficacy / counts.efficacy : 0,
  };
}

File: apps/api/app/services/scoring.py
from app.schemas.assessment import DimensionMeans, DimensionCutoffMap, BurnoutResult


def classify_dimension(
    value: float,
    dimension: str,
    cutoffs: DimensionCutoffMap,
) -> str:
    c = cutoffs[dimension]

    # Efficacy is reverse-scored: lower = higher burnout risk
    if dimension == "efficacy":
        if value >= c["high"]:
            return "low"
        if value <= c["low"]:
            return "high"
        return "medium"

    if value <= c["low"]:
        return "low"
    if value >= c["high"]:
        return "high"
    return "medium"


def compute_overall_risk(
    means: DimensionMeans,
    cutoffs: DimensionCutoffMap,
) -> str:
    levels = [
        classify_dimension(means.exhaustion, "exhaustion", cutoffs),
        classify_dimension(means.cynicism, "cynicism", cutoffs),
        classify_dimension(means.efficacy, "efficacy", cutoffs),
    ]
    highs = levels.count("high")
    if highs >= 2:
        return "high"
    if highs == 1 or "medium" in levels:
        return "medium"
    return "low"


def compute_dimension_means(
    responses: list[dict],
    item_dimensions: dict[int, str],
) -> DimensionMeans:
    sums = {"exhaustion": 0.0, "cynicism": 0.0, "efficacy": 0.0}
    counts = {"exhaustion": 0, "cynicism": 0, "efficacy": 0}

    for r in responses:
        dim = item_dimensions.get(r["item_number"])
        if not dim:
            continue
        sums[dim] += r["value"]
        counts[dim] += 1

    return DimensionMeans(
        exhaustion=sums["exhaustion"] / counts["exhaustion"] if counts["exhaustion"] else 0,
        cynicism=sums["cynicism"] / counts["cynicism"] if counts["cynicism"] else 0,
        efficacy=sums["efficacy"] / counts["efficacy"] if counts["efficacy"] else 0,
    )


def build_burnout_result(
    means: DimensionMeans,
    cutoffs: DimensionCutoffMap,
) -> BurnoutResult:
    return BurnoutResult(
        exhaustion_mean=means.exhaustion,
        cynicism_mean=means.cynicism,
        efficacy_mean=means.efficacy,
        exhaustion_level=classify_dimension(means.exhaustion, "exhaustion", cutoffs),
        cynicism_level=classify_dimension(means.cynicism, "cynicism", cutoffs),
        efficacy_level=classify_dimension(means.efficacy, "efficacy", cutoffs),
        risk_level=compute_overall_risk(means, cutoffs),
    )
